File: app/src/main/java/com/example/bloodbank/VerifyEmailActivity.java
package com.example.bloodbank;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;

import com.google.android.material.button.MaterialButton;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;

public class VerifyEmailActivity extends AppCompatActivity {

    private TextView verifyMsg;
    private TextView verifySubtitle;
    private MaterialButton verifyBtn;
    private FirebaseAuth mAuth;
    private MaterialAlertDialogBuilder progressDialog;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_verify_email);

        initializeViews();
        setupFirebase();
        checkEmailVerification();
    }

    private void initializeViews() {
        verifyMsg = findViewById(R.id.verifyMsg);
        verifySubtitle = findViewById(R.id.verifySubtitle);
        verifyBtn = findViewById(R.id.verifyBtn);
        progressDialog = new MaterialAlertDialogBuilder(this)
                .setMessage("Sending verification email...")
                .setCancelable(false);
    }

    private void setupFirebase() {
        mAuth = FirebaseAuth.getInstance();
    }

    private void checkEmailVerification() {
        FirebaseUser user = mAuth.getCurrentUser();
        if (user != null && !user.isEmailVerified()) {
            showVerificationUI();
            setupVerificationButton();
        } else {
            navigateToMainActivity();
        }
    }

    private void showVerificationUI() {
        verifyMsg.setVisibility(View.VISIBLE);
        verifySubtitle.setVisibility(View.VISIBLE);
        verifyBtn.setVisibility(View.VISIBLE);
    }

    private void setupVerificationButton() {
        verifyBtn.setOnClickListener(v -> {
            FirebaseUser user = mAuth.getCurrentUser();
            if (user != null) {
                sendVerificationEmail(user);
            }
        });
    }

    private void sendVerificationEmail(FirebaseUser user) {
        progressDialog.show();
        
        user.sendEmailVerification()
            .addOnSuccessListener(unused -> {
                progressDialog.create().dismiss();
                Toast.makeText(VerifyEmailActivity.this, 
                    "Verification email has been sent", 
                    Toast.LENGTH_SHORT).show();
                navigateToMainActivity();
            })
            .addOnFailureListener(e -> {
                progressDialog.create().dismiss();
                Toast.makeText(VerifyEmailActivity.this, 
                    "Failed to send verification email: " + e.getMessage(), 
                    Toast.LENGTH_SHORT).show();
            });
    }

    private void navigateToMainActivity() {
        Intent intent = new Intent(VerifyEmailActivity.this, MainActivity.class);
        startActivity(intent);
        finish();
    }
}

File: app/src/main/java/com/example/bloodbank/VerifyEmailActivity.kt
package com.example.bloodbank

import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.TextView
import android.widget.Toast
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.google.android.material.button.MaterialButton
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch

/**
 * VerifyEmailActivity - Email verification screen.
 * Follows MVVM architecture with AuthViewModel.
 */
@AndroidEntryPoint
class VerifyEmailActivity : AppCompatActivity() {
    
    private val viewModel: AuthViewModel by viewModels()
    
    // Views
    private lateinit var verifyMsg: TextView
    private lateinit var verifySubtitle: TextView
    private lateinit var verifyBtn: MaterialButton
    
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_verify_email)
        
        // Initialize views
        initializeViews()
        
        // Check if email is already verified
        checkEmailVerification()
        
        // Setup click listeners
        setupClickListeners()
        
        // Observe ViewModel
        observeViewModel()
    }
    
    private fun initializeViews() {
        verifyMsg = findViewById(R.id.verifyMsg)
        verifySubtitle = findViewById(R.id.verifySubtitle)
        verifyBtn = findViewById(R.id.verifyBtn)
    }
    
    private fun checkEmailVerification() {
        if (viewModel.checkEmailVerification()) {
            navigateToMain()
        } else {
            showVerificationUI()
        }
    }
    
    private fun showVerificationUI() {
        verifyMsg.visibility = View.VISIBLE
        verifySubtitle.visibility = View.VISIBLE
        verifyBtn.visibility = View.VISIBLE
    }
    
    private fun setupClickListeners() {
        verifyBtn.setOnClickListener {
            viewModel.sendEmailVerification()
        }
    }
    
    private fun observeViewModel() {
        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                // Observe authentication state
                launch {
                    viewModel.authState.collect { state ->
                        when (state) {
                            is AuthViewModel.AuthState.EmailVerificationSent -> {
                                Toast.makeText(
                                    this@VerifyEmailActivity,
                                    "Verification email has been sent. Please check your inbox.",
                                    Toast.LENGTH_LONG
                                ).show()
                                navigateToMain()
                            }
                            is AuthViewModel.AuthState.Error -> {
                                // Error is handled by error flow
                            }
                            else -> {
                                // Other states not relevant here
                            }
                        }
                    }
                }
                
                // Observe loading state
                launch {
                    viewModel.isLoading.collect { isLoading ->
                        verifyBtn.isEnabled = !isLoading
                        verifyBtn.text = if (isLoading) "Sending..." else "Send Verification Email"
                    }
                }
                
                // Observe errors
                launch {
                    viewModel.error.collect { error ->
                        error?.let {
                            Toast.makeText(this@VerifyEmailActivity, it, Toast.LENGTH_SHORT).show()
                            viewModel.clearError()
                        }
                    }
                }
            }
        }
    }
    
    private fun navigateToMain() {
        startActivity(Intent(this, MainActivity::class.java))
        finish()
    }
}
